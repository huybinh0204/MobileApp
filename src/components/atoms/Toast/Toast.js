import { useNavigation } from '@react-navigation/native';
import PropTypes from 'prop-types';
import React, { useCallback, useEffect, useState } from 'react';
import { TouchableOpacity } from 'react-native';
import Animated, { Easing, useValue } from 'react-native-reanimated';
import IconSvg from '_components/atoms/IconSvg/IconSvg';
import { ICONS, TOAST_TYPES } from '_constants';
import { normalize } from '_utilities/screen';
import {
  Body,
  CloseIcon,
  Container,
  Header,
  Title,
  ToastCard,
  toastVariants,
} from './Toast.styles';

const Toast = ({
  content,
  duration,
  header,
  offset,
  onClose,
  paddingHorizontal,
  show,
  type,
  ...rest
}) => {
  const { addListener } = useNavigation();
  const animation = useValue(0);
  const [height, setHeight] = useState(0);

  const opacity = show ? 1 : 0;

  const translateY = animation.interpolate({
    inputRange: [0, 1],
    outputRange: [2 * height, -offset],
  });

  const measureToastHeight = ({ nativeEvent }) => {
    setHeight(nativeEvent.layout.height);
  };

  const showToast = useCallback(() => {
    Animated.timing(animation, {
      toValue: 1,
      duration: 200,
      easing: Easing.ease,
    }).start();
  }, [animation]);

  const hideToast = useCallback(() => {
    Animated.timing(animation, {
      toValue: 0,
      duration: 250,
      easing: Easing.in(Easing.ease),
    }).start(({ finished }) => {
      if (finished && onClose) {
        onClose();
      }
    });
  }, [animation, onClose]);

  useEffect(() => {
    let timer;
    let unsubscribe;

    if (show) {
      showToast();
      timer = setTimeout(hideToast, duration);
      unsubscribe = addListener('blur', hideToast);
    }

    return () => {
      clearTimeout(timer);
      unsubscribe?.();
    };
  }, [addListener, duration, hideToast, show, showToast]);

  return (
    <Container
      onLayout={measureToastHeight}
      paddingHorizontal={paddingHorizontal}
      style={{ opacity, transform: [{ translateY }] }}
      {...rest}
    >
      <ToastCard type={type}>
        <Header testID="toastHeader">
          <IconSvg icon={toastVariants[type].icon} width={normalize(20)} height={normalize(20)} />
          <Title numberOfLines={2} type={type}>
            {header}
          </Title>
          <TouchableOpacity
            accessibilityLabel="Close toast"
            accessibilityRole="button"
            testID="toastCloseButton"
            onPress={hideToast}
          >
            <CloseIcon icon={ICONS.closeWhite} type={type} />
          </TouchableOpacity>
        </Header>
        {content ? <Body type={type}>{content}</Body> : null}
      </ToastCard>
    </Container>
  );
};

Toast.defaultProps = {
  content: null,
  header: null,
  duration: 5000,
  offset: normalize(20),
  onClose: null,
  paddingHorizontal: null,
  type: TOAST_TYPES.ERROR,
};

Toast.propTypes = {
  content: PropTypes.string,
  duration: PropTypes.number,
  header: PropTypes.string,
  offset: PropTypes.number,
  onClose: PropTypes.func,
  paddingHorizontal: PropTypes.number,
  show: PropTypes.bool.isRequired,
  type: PropTypes.oneOf(Object.values(TOAST_TYPES)),
};

export default Toast;
