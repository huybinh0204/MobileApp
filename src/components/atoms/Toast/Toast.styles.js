import styled from '@emotion/native';
import Animated from 'react-native-reanimated';
import BaseText from '_components/atoms/BaseText/BaseText';
import Card from '_components/atoms/Card/Card';
import { COLORS, ICONS } from '_constants';
import { normalize } from '_utilities/screen';
import IconSvg from '../IconSvg/IconSvg';

export const Container = styled(Animated.View)(({ paddingHorizontal, theme }) => ({
  position: 'absolute',
  bottom: 0,
  width: '100%',
  alignItems: 'center',
  paddingHorizontal: paddingHorizontal ?? theme.spacing.m,
}));

export const toastVariants = {
  error: {
    background: COLORS.epsilon500,
    color: COLORS.epsilon500,
    closeIcon: ICONS.closeError,
    icon: ICONS.circledExclamation,
  },
  success: {
    background: COLORS.successGreen1,
    color: COLORS.successGreen1,
    closeIcon: ICONS.closeSuccess,
    icon: ICONS.circledCheckmark,
  },
  alert: {
    background: COLORS.zeta800,
    color: COLORS.zeta800,
    closeIcon: ICONS.closeAlert,
    icon: ICONS.toastAlert,
  },
  info: {
    background: COLORS.alpha500,
    color: COLORS.alpha500,
    closeIcon: ICONS.closeInfo,
    icon: ICONS.toastInfo,
  },
};

export const ToastCard = styled(Card)(({ type }) => ({
  width: '100%',
  borderWidth: 1,
  borderColor: toastVariants[type].color,
  backgroundColor: toastVariants[type].background,
}));

export const Header = styled.View({
  flexDirection: 'row',
  alignItems: 'flex-start',
  justifyContent: 'space-between',
});

export const Title = styled(BaseText)(({ theme, type }) => ({
  ...theme.typography.tinyContentBold,
  color: theme.colors.white,
  flex: 1,
  paddingHorizontal: normalize(10),
  paddingTop: 2,
}));

export const Body = styled(BaseText)(({ theme, type }) => ({
  ...theme.typography.tinyContentRegular,
  color: theme.colors.white,
  paddingHorizontal: normalize(30),
  paddingTop: 3,
}));

export const CloseIcon = styled(IconSvg)(({ theme, type }) => ({
  width: normalize(theme.spacing.general),
  height: normalize(theme.spacing.general),
}));
