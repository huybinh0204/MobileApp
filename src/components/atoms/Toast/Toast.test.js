import { render } from '@testing-library/react-native';
import React from 'react';
import ComponentWithProviders from '_test/mocks/ComponentWithProviders';
import Toast from './Toast';

describe('Toast', () => {
  let props;

  beforeEach(() => {
    props = {
      content: 'content',
      header: 'header',
      show: true,
    };
  });

  it('should match the snapshot', () => {
    const { toJSON } = render(<ComponentWithProviders component={Toast} componentProps={props} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render header and body', () => {
    const { getByText } = render(
      <ComponentWithProviders component={Toast} componentProps={props} />
    );

    expect(getByText('header')).toBeTruthy();
    expect(getByText('content')).toBeTruthy();
  });
});
