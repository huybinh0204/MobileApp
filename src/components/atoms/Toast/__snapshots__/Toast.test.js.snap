// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast should match the snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        onLayout={[Function]}
                        paddingHorizontal={null}
                        style={
                          Array [
                            Object {
                              "alignItems": "center",
                              "bottom": 0,
                              "paddingHorizontal": 58,
                              "position": "absolute",
                              "width": "100%",
                            },
                            Object {
                              "opacity": 1,
                              "transform": Array [
                                Object {
                                  "translateY": AnimatedValue {
                                    " __value": 0,
                                  },
                                },
                              ],
                            },
                          ]
                        }
                      >
                        <View
                          style={
                            Array [
                              Object {
                                "alignSelf": "stretch",
                                "backgroundColor": "#FFFFFF",
                                "borderColor": "#F2F3F4",
                                "borderRadius": 10,
                                "borderWidth": 0,
                                "padding": 29,
                                "shadowColor": "#362E60",
                                "shadowOffset": Object {
                                  "height": 7,
                                  "width": 0,
                                },
                                "shadowOpacity": 0.1,
                                "shadowRadius": 10,
                              },
                              Array [
                                Object {
                                  "backgroundColor": "#D02B35",
                                  "borderColor": "#D02B35",
                                  "borderWidth": 1,
                                  "width": "100%",
                                },
                                undefined,
                              ],
                            ]
                          }
                          type="error"
                        >
                          <View
                            style={
                              Array [
                                Object {
                                  "alignItems": "flex-start",
                                  "flexDirection": "row",
                                  "justifyContent": "space-between",
                                },
                                undefined,
                              ]
                            }
                            testID="toastHeader"
                          >
                            <SvgMock
                              height={36}
                              width={36}
                            />
                            <Text
                              allowFontScaling={false}
                              numberOfLines={2}
                              style={
                                Array [
                                  Object {
                                    "color": "#FFFFFF",
                                    "flex": 1,
                                    "fontFamily": "Poppins-SemiBold",
                                    "fontSize": 21.5,
                                    "lineHeight": 30.5,
                                    "paddingHorizontal": 18,
                                    "paddingTop": 2,
                                  },
                                  undefined,
                                ]
                              }
                              type="error"
                            >
                              header
                            </Text>
                            <View
                              accessibilityLabel="Close toast"
                              accessibilityRole="button"
                              accessible={true}
                              collapsable={false}
                              focusable={true}
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Object {
                                  "opacity": 1,
                                }
                              }
                              testID="toastCloseButton"
                            >
                              <SvgMock
                                height={20}
                                style={
                                  Array [
                                    Object {
                                      "height": 65,
                                      "width": 65,
                                    },
                                    undefined,
                                  ]
                                }
                                type="error"
                                width={20}
                              />
                            </View>
                          </View>
                          <Text
                            allowFontScaling={false}
                            style={
                              Array [
                                Object {
                                  "color": "#FFFFFF",
                                  "fontFamily": "Poppins",
                                  "fontSize": 21.5,
                                  "lineHeight": 30.5,
                                  "paddingHorizontal": 54.5,
                                  "paddingTop": 3,
                                },
                                undefined,
                              ]
                            }
                            type="error"
                          >
                            content
                          </Text>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "height": 44,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1,
          }
        }
      />
    </View>
  </RNCSafeAreaProvider>
</View>
`;
